actor/root_context.go

```diff
@@ -0,0 +1,116 @@
+package actor
+
+import "time"
+
+type RootContext struct {
+	senderMiddleware SenderFunc
+	headers          messageHeader
+}
+
+var emptyRootContext = &RootContext{
+	senderMiddleware: nil,
+	headers:          emptyMessageHeader,
+}
+
+func EmptyRootContext() *RootContext {
+	return emptyRootContext
+}
+
+func NewRootContext(header map[string]string, middleware ...SenderMiddleware) *RootContext {
+	return &RootContext{
+		senderMiddleware: makeSenderMiddlewareChain(middleware, func(_ SenderContext, target *PID, envelope *MessageEnvelope) {
+			target.sendUserMessage(envelope)
+		}),
+		headers: messageHeader(header),
+	}
+}
+
+func (rc *RootContext) WithHeaders(headers map[string]string) *RootContext {
+	rc.headers = headers
+	return rc
+}
+
+func (rc *RootContext) WithSenderMiddleware(middleware ...SenderMiddleware) *RootContext {
+	rc.senderMiddleware = makeSenderMiddlewareChain(middleware, func(_ SenderContext, target *PID, envelope *MessageEnvelope) {
+		target.sendUserMessage(envelope)
+	})
+	return rc
+}
+
+//
+// Interface: SenderContext
+//
+
+func (rc *RootContext) Message() interface{} {
+	return nil
+}
+
+func (rc *RootContext) MessageHeader() ReadonlyMessageHeader {
+	return rc.headers
+}
+
+func (rc *RootContext) Send(pid *PID, message interface{}) {
+	rc.sendUserMessage(pid, message)
+}
+
+func (rc *RootContext) Request(pid *PID, message interface{}) {
+	rc.sendUserMessage(pid, message)
+}
+
+func (rc *RootContext) RequestWithCustomSender(pid *PID, message interface{}, sender *PID) {
+	env := &MessageEnvelope{
+		Header:  nil,
+		Message: message,
+		Sender:  sender,
+	}
+	rc.sendUserMessage(pid, env)
+}
+
+// RequestFuture sends a message to a given PID and returns a Future
+func (rc *RootContext) RequestFuture(pid *PID, message interface{}, timeout time.Duration) *Future {
+	future := NewFuture(timeout)
+	env := &MessageEnvelope{
+		Header:  nil,
+		Message: message,
+		Sender:  future.PID(),
+	}
+	rc.sendUserMessage(pid, env)
+	return future
+}
+
+func (rc *RootContext) sendUserMessage(pid *PID, message interface{}) {
+	if rc.senderMiddleware != nil {
+		if envelope, ok := message.(*MessageEnvelope); ok {
+			// Request based middleware
+			rc.senderMiddleware(rc, pid, envelope)
+		} else {
+			// tell based middleware
+			rc.senderMiddleware(rc, pid, &MessageEnvelope{nil, message, nil})
+		}
+		return
+	}
+	// Default path
+	pid.sendUserMessage(message)
+}
+
+//
+// Interface: SpawnerContext
+//
+
+func (rc *RootContext) Spawn(props *Props) (*PID, error) {
+	name := ProcessRegistry.NextId()
+	return rc.SpawnNamed(props, name)
+}
+
+func (rc *RootContext) SpawnPrefix(props *Props, prefix string) (*PID, error) {
```

Comment:
The signature on this should be

```suggestion
func (rc *RootContext) SpawnPrefix(props *Props, prefix string) *PID {
```

Reply:
this! I wanted to contribute that exact piece but I would prefer to get this PR forward.

Reply:
@emilingerslev  @Chumper  Hmm, sorry but I wasn't able to put myself into the shoes for this request. Could you guys describe a bit more about why the error should be removed? Thank you very much!  = )

Reply:
We have two methods for spawning an actor, either in the `root_context` or in the `context` of an actor.

In the context of an actor we do not return an error but just the `PID`:
https://github.com/AsynkronIT/protoactor-go/blob/match-dotnet/actor/context.go#L102

So it makes sense to match the returns for also spawning in the root context.

My reason for removing: As that will spawn an actor with a prefixed name and a unique id that is taken from the process registry there should be no case where the name is already taken.

If that is the case I would opt for a panic inside the method because then something else is wrong in the system.

Reply:
Ok, there could be edge cases where the `SpawnPrefix` method could return the same name as an already existing actor:
```
context.SpawnNamed(props, "Foo$2")
context.SpawnPrefix(props, "Foo")
```
or similar shenanigans, the question is do we want to catch that? If so, then we should add the error to all spawn methods.

Reply:
@Chumper Thank you very much for the explanation! I opted in for panic option since calling SpawnNamed with a name like "Foo$2" should always be avoided, I just added this in documentation. This has been fixed in latest commits.